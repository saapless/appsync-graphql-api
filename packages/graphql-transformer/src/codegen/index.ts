export * from "./ast";
export * from "./printer";
export * from "./document";
