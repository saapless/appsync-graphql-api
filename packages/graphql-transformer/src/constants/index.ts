export const ConnectionDirective = Object.freeze({
  HAS_ONE: "hasOne",
  HAS_MANY: "hasMany",
});

export const ModelDirective = Object.freeze({
  MODEL: "model",
});

export const UtilityDirective = Object.freeze({
  SERVER_ONLY: "serverOnly",
  CLIENT_ONLY: "clientOnly",
  READ_ONLY: "readOnly",
  WRITE_ONLY: "writeOnly",
  FILTER_ONLY: "filterOnly",
  INTERNAL: "internal",
});

export const ScalarType = Object.freeze({
  STRING: "String",
  INT: "Int",
  FLOAT: "Float",
  BOOLEAN: "Boolean",
  ID: "ID",
  LONG: "Long",
});

export const RelationType = Object.freeze({
  ONE_TO_ONE: "oneToOne",
  ONE_TO_MANY: "oneToMany",
  MANY_TO_MANY: "manyToMany",
});
