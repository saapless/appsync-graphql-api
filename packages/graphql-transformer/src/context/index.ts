export * from "./TransformerContext";
export type {
  DescriptorOperation,
  Key,
  KeyOperator,
  KeyValue,
  OperationActionType,
  ResolverDescriptor,
  ReturnType,
} from "./ResolverLoader";
