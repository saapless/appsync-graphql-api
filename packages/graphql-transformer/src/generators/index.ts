export { ResolverGeneratorBase } from "./ResolverGeneratorBase";
export { DynamoDbGenerator } from "./DynamoDBGenerator";
export { NoneGenerator } from "./NoneGenerator";
export { SchemaTypesGenerator } from "./SchemaTypesGenerator";
export { ResolverTypesGenerator } from "./ResolverTypesGenerator";
