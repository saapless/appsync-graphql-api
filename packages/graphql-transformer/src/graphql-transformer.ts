import { GraphQLTransformer } from "./GraphQLTransformer";

export function createTransformer() {
  return new GraphQLTransformer();
}

export function executeTranformer() {}
