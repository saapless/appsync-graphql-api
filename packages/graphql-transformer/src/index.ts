export * from "./transformer";
export { definitionFromSource } from "./utils";
