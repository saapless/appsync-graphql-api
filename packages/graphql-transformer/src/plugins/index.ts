export { TransformerPluginBase, type IPluginFactory } from "./TransformerPluginBase";
export { AWSTypesPlugin } from "./AWSTypesPlugin";
export { AuthPlugin } from "./AuthPlugin";
export { ModelPlugin } from "./ModelPlugin";
export { NodeInterfacePlugin } from "./NodeInterfacePlugin";
export { ConnectionPlugin } from "./ConnectionPlugin";
