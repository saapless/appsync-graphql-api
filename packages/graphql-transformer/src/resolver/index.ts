export { ResolverBase } from "./ResolverBase";
export { FieldResolver } from "./FieldResolver";
export { FunctionResolver } from "./FunctionResolver";
