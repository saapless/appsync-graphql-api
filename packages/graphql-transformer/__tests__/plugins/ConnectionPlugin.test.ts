import { TransformerContext } from "../../src";
import {
  DirectiveDefinitionNode,
  DocumentNode,
  EnumNode,
  FieldNode,
  InputObjectNode,
  ObjectNode,
} from "../../src/parser";
import { ConnectionPlugin } from "../../src/plugins";
import { FieldResolver } from "../../src/resolver";

const schema = /* GraphQL */ `
  type User {
    id: ID!
    username: String!
  }

  type Todo {
    id: ID
    content: String
    resources: Resource @edges
  }

  type Document {
    url: String!
  }

  type Message {
    content: String!
  }

  union Resource = Document | Message

  type Query {
    me: User @node(key: "id", ref: "identity.sub")
    todos: Todo
      @connection(key: "userId", ref: "identity.sub", sk: { beginsWith: "Todo" }, index: "byUserId")
  }
`;

describe("ConnectionPlugin", () => {
  const context = new TransformerContext({ document: DocumentNode.fromSource(schema) });
  const plugin = ConnectionPlugin.create(context);

  describe("on run `before` hook", () => {
    beforeAll(() => plugin.before());

    it("adds connection directive definitions", () => {
      expect(context.document.getNode("ConnectionRelationType")).toBeInstanceOf(EnumNode);
      expect(context.document.getNode("SortKeyInput")).toBeInstanceOf(InputObjectNode);
      expect(context.document.getNode("connection")).toBeInstanceOf(DirectiveDefinitionNode);
      expect(context.document.getNode("edges")).toBeInstanceOf(DirectiveDefinitionNode);
      expect(context.document.getNode("node")).toBeInstanceOf(DirectiveDefinitionNode);
    });
  });

  describe("on normalize node", () => {
    beforeAll(() => {
      plugin.normalize(context.document.getQueryNode());
      plugin.normalize(context.document.getNode("Todo") as ObjectNode);
    });

    it("updates `node` connection arguments", () => {
      const queryNode = context.document.getQueryNode();
      expect(queryNode.getField("me")?.hasDirective("node")).toBeFalsy();
      expect(queryNode.getField("me")?.hasDirective("connection")).toBeTruthy();
    });

    it("updates `edges` connection arguments", () => {
      const todoNode = context.document.getNode("Todo") as ObjectNode;
      expect(todoNode.getField("resources")?.hasDirective("edges")).toBeFalsy();
      expect(todoNode.getField("resources")?.hasDirective("connection")).toBeTruthy();
    });

    it("adds connection keys to nodes", () => {
      expect((context.document.getNode("Todo") as ObjectNode).getField("userId")).toBeInstanceOf(
        FieldNode
      );
      expect(
        (context.document.getNode("Document") as ObjectNode).getField("sourceId")
      ).toBeInstanceOf(FieldNode);
      expect(
        (context.document.getNode("Message") as ObjectNode).getField("sourceId")
      ).toBeInstanceOf(FieldNode);
    });
  });

  describe("on execute object node", () => {
    beforeAll(() => {
      plugin.execute(context.document.getNode("Query") as ObjectNode);
      plugin.execute(context.document.getNode("Todo") as ObjectNode);
    });

    it("creates connection types", () => {
      expect(context.document.getNode("TodoConnection")).toBeInstanceOf(ObjectNode);
      expect(context.document.getNode("TodoEdge")).toBeInstanceOf(ObjectNode);
      expect(context.document.getNode("ResourceConnection")).toBeInstanceOf(ObjectNode);
      expect(context.document.getNode("ResourceEdge")).toBeInstanceOf(ObjectNode);
    });

    it("updates field types", () => {
      const todosField = (context.document.getNode("Query") as ObjectNode).getField("todos");
      const resourcesField = (context.document.getNode("Todo") as ObjectNode).getField("resources");

      expect(todosField?.type.getTypeName()).toBe("TodoConnection");
      expect(todosField?.getArgument("filter")?.type.getTypeName()).toBe("TodoFilterInput");
      expect(todosField?.hasArgument("first")).toBe(true);
      expect(todosField?.hasArgument("after")).toBe(true);
      expect(todosField?.hasArgument("sort")).toBe(true);

      expect(resourcesField?.type.getTypeName()).toBe("ResourceConnection");
      expect(resourcesField?.getArgument("filter")?.type.getTypeName()).toBe("ResourceFilterInput");
      expect(resourcesField?.hasArgument("first")).toBe(true);
      expect(resourcesField?.hasArgument("after")).toBe(true);
      expect(resourcesField?.hasArgument("sort")).toBe(true);
    });

    it("creates field resolvers", () => {
      expect(context.resolvers.get("Todo.resources")).toBeInstanceOf(FieldResolver);
    });
  });

  describe("on cleanup node", () => {
    beforeAll(() => {
      plugin.cleanup(context.document.getNode("Query") as ObjectNode);
      plugin.cleanup(context.document.getNode("Todo") as ObjectNode);
    });

    it("removes connection directives from fields", () => {
      const queryNode = context.document.getNode("Query") as ObjectNode;
      const todoNode = context.document.getNode("Todo") as ObjectNode;

      expect(queryNode.getField("me")?.hasDirective("connection")).toBeFalsy();
      expect(queryNode.getField("todos")?.hasDirective("connection")).toBeFalsy();
      expect(todoNode.getField("resources")?.hasDirective("connection")).toBeFalsy();
    });
  });

  describe("on run `after` hook", () => {
    beforeAll(() => {
      plugin.after();
    });

    it("removes connection directive definitions", () => {
      expect(context.document.getNode("ConnectionRelationType")).toBeUndefined();
      expect(context.document.getNode("SortKeyInput")).toBeUndefined();
      expect(context.document.getNode("connection")).toBeUndefined();
      expect(context.document.getNode("edges")).toBeUndefined();
      expect(context.document.getNode("node")).toBeUndefined();
    });
  });

  it("matches snapshot", () => {
    expect(context.document.print()).toMatchSnapshot();
  });
});
