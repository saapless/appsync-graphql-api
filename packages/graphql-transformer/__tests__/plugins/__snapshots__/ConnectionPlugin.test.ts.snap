// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionPlugin matches snapshot 1`] = `
"type User {
  id: ID!
  username: String!
}

type Todo {
  id: ID
  content: String
  resources(filter: ResourceFilterInput, first: Int, after: String, sort: ModelSortDirection): ResourceConnection!
  userId: ID @private
}

type Document {
  url: String!
  sourceId: ID @writeonly
}

type Message {
  content: String!
  sourceId: ID @writeonly
}

union Resource = Document | Message

type Query {
  me: User
  todos(filter: TodoFilterInput, first: Int, after: String, sort: ModelSortDirection): TodoConnection!
}

type PageInfo {
  hasNextPage: Boolean
  hasPreviousPage: Boolean
  startCursor: String
  endCursor: String
}

type TodoConnection {
  edges: [TodoEdge!]!
  pageInfo: PageInfo!
}

type TodoEdge {
  cursor: String
  node: Todo
}

input TodoFilterInput {
  id: ModelIDInput
  content: ModelStringInput
  and: [TodoFilterInput]
  or: [TodoFilterInput]
  not: TodoFilterInput
}

type ResourceConnection {
  edges: [ResourceEdge!]!
  pageInfo: PageInfo!
}

type ResourceEdge {
  cursor: String
  node: Resource
}

input ResourceFilterInput {
  __typename: ModelStringInput
  url: ModelStringInput
  content: ModelStringInput
  and: [ResourceFilterInput]
  or: [ResourceFilterInput]
  not: ResourceFilterInput
}"
`;
